File: 2024/01/com/aoc/Day01.java
package com.aoc;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

public class Day01 implements Day{
    private final InputHelper inputHelper;
    ArrayList<String> lines;

    private static final String DAY = "01";

    private ArrayList<Integer> list1;
    private ArrayList<Integer> list2;

    public Day01(boolean testing){
        this.inputHelper = new InputHelper();
        this.lines = this.inputHelper.readLines(testing, DAY);
        initInput();
    }

    private void initInput(){
        this.list1 = new ArrayList<Integer>();
        this.list2 = new ArrayList<Integer>();

        for (String line : lines){
            String[] l = line.split(" {3}");
            list1.add(Integer.valueOf(l[0]));
            list2.add(Integer.valueOf(l[1]));
        }
    }

    @Override
    public void solveStage1() {
        System.out.println("Solving Stage 1...");
        Collections.sort(this.list1);
        Collections.sort(this.list2);

        int difference = 0;
        for (int i = 0; i < list1.size(); i++){
            difference += Math.abs(list1.get(i) - list2.get(i));
        }

        System.out.println("Solution Stage 1: " + difference);
    }

    @Override
    public void solveStage2() {
        System.out.println("Solving Stage 2...");
        HashMap<Integer, Integer> countMap = new HashMap<Integer, Integer>();
        for(Integer id : list2){
            countMap.compute(id , (k, count) -> (count == null) ? 1 : (count + 1));
        }

        int similarity = 0;
        for(Integer id : list1){
            similarity += (id * countMap.getOrDefault(id, 0));
        }

        System.out.println("Solution Stage 2: " + similarity);
    }
}

File: 2025/01/Day01.cs
using AdventOfCode.helper;

namespace AdventOfCode._01;

public class Day01 : Day
{
    private InputHelper helper;
    private bool test;

    private List<string> lines;
    
    private static string DAY = "01";

    public Day01(bool test)
    {
        this.test = test;
        this.helper = new InputHelper();
        this.lines = helper.ReadLines(this.test, DAY);
        if (this.lines.Count == 0) {
            System.Environment.Exit(-1);
        }
        
    }
    public void solveStage1()
    {
        var dial = 50;
        var count = 0;
        this.lines.ForEach(line =>
        {
            int n = Int32.Parse(line.Substring(1));

            if (line[0] == 'L')
                dial -= n;
            else 
                dial += n;

            count += dial % 100 == 0 ? 1 : 0;
            
        });

        Console.WriteLine($"Solution stage 1: {count}");
    }

    public void solveStage2()
    {
        var dial = 50;
        var count = 0;
        this.lines.ForEach(line =>
        {
            int n = Int32.Parse(line.Substring(1));
            var newDial = dial;
            var hundreds = n / 100;
            n -= 100 * hundreds;
            count += hundreds;

            
            if (line[0] == 'L')
            {
                newDial = (dial + 100 - n) % 100;
                count += ((dial % 100 != 0) && newDial > dial) || (newDial % 100 == 0) ? 1 : 0;
            }
            else
            {
                newDial = (dial + 100+n) % 100;
                count += ((dial % 100 != 0) && newDial < dial) || (newDial % 100 == 0) ? 1 : 0;
            }
            dial = newDial;
        });

        Console.WriteLine($"Solution stage 2: {count}");
    }
}